### PR TITLE
Rishabh Malviya Module 1

### PR DESCRIPTION
**Pull request by rishu541**
Completed Module 1
### Things I have worked on
- I have worked on **git** and **GitHub**.
- I have learned about java concepts.
- I have implemented different **Data structures**.
- I have learned about different **collision** handling techniques.

Thanks, mentors for clearing my doubts

### CODE REVIEW

File: Data Structures/Graph.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        public void addEdge(int src, int dest){
+            Node newNode = new Node(dest);
+            newNode.next = adjList[src];
+            adjList[src] = newNode;
+            newNode = new Node(src);
+            newNode.next = adjList[dest];
+            adjList[dest] = newNode;
+        }
+        public void printGraph(){
+            for(int i = 0; i < V; i++){
```
Comment (Rohit-pawar902) [inline Data Structures/Graph.java]: give purpose-full variable name on iterations also

Comment (Rohit-pawar902) [reply Data Structures/Graph.java]: And for temp named variable ,find its purpose what is scenario and then give another name

File: Data Structures/HashMap.java [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        hashMap(int size) {
+            this.size = size;
+            arr = new Pair[size];
+            for (int i = 0; i < size; i++) {
+                arr[i] = null;
+            }
+        }
+
+        //  Function to insert a key value pair in the hash map
+        public int hashFunction(int key) {
```
Comment (Rohit-pawar902) [inline Data Structures/HashMap.java:26]: Can we use Better Hash-Function here or not?

File: Data Structures/HashMap.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+            while (temp != null) {
+                if (temp.key == key) {
+                    return temp.value;
+                }
+                temp = temp.next;
+            }
+            return -1;
+        }
+
+        public void display() {
```
Comment (Rohit-pawar902) [inline Data Structures/HashMap.java:75]: purpose-full-variable name ,you have to conscious about it.

File: Data Structures/Stack.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+               return 0;
+           }
+           else{
+               int data = head.data;
+               head = head.next;
+               return data;
+           }
+        }
+//        Method to update an element in the Stack
+        private static void updateElement(Node head,int nextInt){
```
Comment (Rohit-pawar902) [inline Data Structures/Stack.java:112]: Is stack gives this option for update on particular element?

File: Data Structures/linkedList.java [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+            fast = fast.next;
+            if(fast != null){
+                slow = slow.next;
+                fast = fast.next;
+            }
+        }
+        return slow;
+    }
+    //    Function to merge two sorted linked lists
+    static Node merge(Node a,Node b){
```
Comment (Rohit-pawar902) [inline Data Structures/linkedList.java:163]: Variable name.

File: Hashing/DoubleHashing.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        elements = 0;
+    }
+    boolean isFull(){
+        return elements == size;
+    }
+    int hashFunction(int key){
+        return key%size;
+    }
+    int hasFunction2(int key){
+        return 7-(key%7);
```
Comment (Rohit-pawar902) [inline Hashing/DoubleHashing.java]: Why would you use 7 here ,can we use any other number ? explain about it.

File: Data Structures/Graph.java (hunk 1 above)
Comment (rishu541) [reply Data Structures/Graph.java]: Okay, sir, I will update the variable names to meaningful names.

File: Data Structures/HashMap.java (hunk 2 above)
Comment (rishu541) [reply Data Structures/HashMap.java:26]: Yes, Sir, we can use a better hash function to generate a key.
by using random numbers, timestamps, or prime numbers.

File: Data Structures/Stack.java (hunk 4 above)
Comment (rishu541) [reply Data Structures/Stack.java:112]: No Sir, I will change it.

File: Data Structures/linkedList.java (hunk 5 above)
Comment (rishu541) [reply Data Structures/linkedList.java:163]: Okay sir, I will change to meaningful name

File: Hashing/DoubleHashing.java (hunk 6 above)
Comment (rishu541) [reply Hashing/DoubleHashing.java]: Yes, I can use prime number which tends to generate a unique hash every time.
I have used hardcoded `7` (prime number). I can write a get prime function to get a prime number every time.

File: Data Structures/HashMap.java (hunk 3 above)
Comment (rishu541) [reply Data Structures/HashMap.java:75]: okay sir

File: Hashing/RandomProbing.java [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+    void display() {
+        for (int i = 0; i < capacity; i++) {
+            if (table[i] != null) {
+                System.out.println("Key: " + table[i].key + " Value: " + table[i].value);
+            }
+        }
+    }
+
+    int hash(int key) {
```
Comment (Rohit-pawar902) [inline Hashing/RandomProbing.java:33]: Implement random probing in right way.

**Pull request closed without merging**